pyccel/codegen/wrapper/fortran_to_c_wrapper.py

```diff
@@ -315,59 +263,149 @@ def _wrap_Interface(self, expr):
+                func_def_argument_dict['f_arg'] = FunctionCallArgument(func_def_argument_dict['f_arg'],
+                                                                        keyword = expr.name)
+                return func_def_argument_dict
+
+        # Unknown object, we raise an error.
+        return errors.report(f"Wrapping function arguments is not implemented for type {class_type}. "
+                + PYCCEL_RESTRICTION_TODO, symbol=var, severity='fatal')
+
+    def _extract_FixedSizeNumericType_FunctionDefArgument(self, var, func):
         name = var.name
         self.scope.insert_symbol(name)
         collisionless_name = self.scope.get_expected_name(var.name)
-        if isinstance(var.class_type, (NumpyNDArrayType, HomogeneousTupleType)) or \
-                var.is_optional or isinstance(var.dtype, CustomDataType):
+        if var.is_optional:
+            f_arg = var.clone(collisionless_name, new_class = Variable, is_argument = False,
+                    is_optional = False, memory_handling='alias')
             new_var = Variable(BindCPointer(), self.scope.get_new_name(f'bound_{name}'),
                                 is_argument = True, is_optional = False, memory_handling='alias')
-            arg_var = var.clone(collisionless_name, is_argument = False, is_optional = False,
-                                memory_handling = 'alias', allows_negative_indexes=False,
-                                new_class = Variable)
-            self.scope.insert_variable(arg_var)
+            body = [C_F_Pointer(new_var, f_arg)]
         else:
-            new_var = var.clone(collisionless_name, new_class = Variable)
-        self.scope.insert_variable(new_var)
+            f_arg = var.clone(collisionless_name, new_class = Variable,
+                        is_argument = True)
+            new_var = f_arg
+            body = []
+        self.scope.insert_variable(f_arg)
+        return {'c_arg': BindCVariable(new_var, var), 'f_arg': f_arg, 'body': body}
+
+    def _extract_CustomDataType_FunctionDefArgument(self, var, func):
+        name = var.name
+        self.scope.insert_symbol(name)
+        collisionless_name = self.scope.get_expected_name(var.name)
+        f_arg = var.clone(collisionless_name, new_class = Variable, is_argument = False,
+                is_optional = False, memory_handling='alias')
+        new_var = Variable(BindCPointer(), self.scope.get_new_name(f'bound_{name}'),
+                            is_argument = True, is_optional = False, memory_handling='alias')
+        body = [C_F_Pointer(new_var, f_arg)]
+        self.scope.insert_variable(f_arg)
+        return {'c_arg': BindCVariable(new_var, var), 'f_arg': f_arg, 'body': body}
+
+    def _extract_NumpyNDArrayType_FunctionDefArgument(self, var, func):
+        name = var.name
+        scope = self.scope
+        scope.insert_symbol(name)
+        collisionless_name = scope.get_expected_name(var.name)
```

Comment:
`var.name` can be replaced with `name`

Reply:
Fixed [f9365af](https://github.com/pyccel/pyccel/pull/2232/commits/f9365af03c0482c7585c3d4288b8df0b6307a067)